### BY OFURE FORTUNATE
                                                       November 2023

# Analyzing Hotel Booking Data with Python to Solve Guests Retention Problem.

## Project Overview & Objectives :



An elite hotel chain managing 2 hotels(City and Resort Hotels), aims to enhance guests retention by understanding the factors contributing to

- booking cancellations, 

- short stays, and 

- low customer return rates.

As a Data Analyst Consultant for this hotel chain, my task is to analyze the hotel’s booking data and devise strategies that improve guest satisfaction, reduce cancellations, increase the length of stays, and encourage repeat bookings.


## Data Source

The dataset was sourced from Kaggle and originally featured in the article "Hotel Booking Demand Datasets," authored by Nuno Antonio, Ana Almeida, and Luis Nunes for Data in Brief, Volume 22, February 2019. 
It originally had 36 columns and 119,390 booking observations for both a City Hotel and a Resort Hotel.

The dataset includes a range of information, such as guests’ personal details (names, email, 
phone-numbers, country, payment information), relevant dates like arrival & reservation status dates, and booking specifics like the number of nights spent, meal, reserved room type, deposit type, and required car parking spaces. There are also essential details such as hotel, whether the customer canceled, lead time, market segment, distribution channel, customer type, ADR (average daily rate), days in the waiting list, previous cancellations, and whether they are repeated guests.

I conducted data cleaning on the dataset to allow for seamless and accurate analysis. 

(**Refer to "HOTEL BOOKING DATA CLEANING" file for the process.**)

In [140]:
#Importing relevant libraries and the dataset

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [141]:
df = pd.read_csv("C:\\Users\\840 g5\\Downloads\\hotel_data.csv")

## **Preliminary Exploratory Data Analysis & DataSet Transformation**



In [142]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118898 entries, 0 to 118897
Data columns (total 35 columns):
 #   Column                          Non-Null Count   Dtype 
---  ------                          --------------   ----- 
 0   Unnamed: 0                      118898 non-null  int64 
 1   name                            118898 non-null  object
 2   email                           118898 non-null  object
 3   phone-number                    118898 non-null  object
 4   credit_card                     118898 non-null  object
 5   arrival_date                    118898 non-null  object
 6   reservation_status_date         118898 non-null  object
 7   reservation_status              118898 non-null  object
 8   hotel                           118898 non-null  object
 9   is_canceled                     118898 non-null  bool  
 10  lead_time                       118898 non-null  int64 
 11  arrival_date_year               118898 non-null  int64 
 12  arrival_date_month            

In [143]:
df.describe()

Unnamed: 0      lead_time  arrival_date_year  \
count  118898.000000  118898.000000      118898.000000   
mean    59872.784496     104.311435        2016.157656   
std     34418.708956     106.903309           0.707459   
min         0.000000       0.000000        2015.000000   
25%     30185.250000      18.000000        2016.000000   
50%     59930.500000      69.000000        2016.000000   
75%     89663.750000     161.000000        2017.000000   
max    119389.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             118898.000000              118898.000000   
mean                  27.166555                  15.800880   
std                   13.589971                   8.780324   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       num_of_weekend_nights  num_of_week_nights         adults  \
count          118898.000000       118898.000000  118898.000000   
mean                0.928897            2.502145       1.858391   
std                 0.996216            1.900168       0.578576   
min                 0.000000            0.000000       0.000000   
25%                 0.000000            1.000000       2.000000   
50%                 1.000000            2.000000       2.000000   
75%                 2.000000            3.000000       2.000000   
max                16.000000           41.000000      55.000000   

            children         babies  previous_cancellations  \
count  118898.000000  118898.000000           118898.000000   
mean        0.104207       0.007948                0.087142   
std         0.399172       0.097380                0.845869   
min         0.000000       0.000000                0.000000   
25%         0.000000       0.000000                0.000000   
50%         0.000000       0.000000                0.000000   
75%         0.000000       0.000000                0.000000   
max        10.000000      10.000000               26.000000   

       previous_bookings_not_canceled  booking_changes  days_in_waiting_list  \
count                   118898.000000    118898.000000         118898.000000   
mean                         0.131634         0.221181              2.330754   
std                          1.484672         0.652785             17.630452   
min                          0.000000         0.000000              0.000000   
25%                          0.000000         0.000000              0.000000   
50%                          0.000000         0.000000              0.000000   
75%                          0.000000         0.000000              0.000000   
max                         72.000000        21.000000            391.000000   

       required_car_parking_spaces  
count                118898.000000  
mean                      0.061885  
std                       0.244172  
min                       0.000000  
25%                       0.000000  
50%                       0.000000  
75%                       0.000000  
max                       8.000000

In [144]:
df

Unnamed: 0                name                        email  \
0                0       Ernest Barnes  Ernest.Barnes31@outlook.com   
1                1        Andrea Baker       Andrea_Baker94@aol.com   
2                2      Rebecca Parker   Rebecca_Parker@comcast.net   
3                3        Laura Murray            Laura_M@gmail.com   
4                4         Linda Hines           LHines@verizon.com   
...            ...                 ...                          ...   
118893      119385     Claudia Johnson          Claudia.J@yahoo.com   
118894      119386      Wesley Aguilar         WAguilar@xfinity.com   
118895      119387        Mary Morales     Mary_Morales@hotmail.com   
118896      119388  Caroline Conley MD      MD_Caroline@comcast.net   
118897      119389      Ariana Michael         Ariana_M@xfinity.com   

        phone-number       credit_card arrival_date reservation_status_date  \
0       669-792-1661  ************4322   2015-07-01              2015-07-01   
1       858-637-6955  ************9157   2015-07-01              2015-07-01   
2       652-885-2745  ************3734   2015-07-01              2015-07-02   
3       364-656-8427  ************5677   2015-07-01              2015-07-02   
4       713-226-5883  ************5498   2015-07-01              2015-07-03   
...              ...               ...          ...                     ...   
118893  403-092-5582  ************8647   2017-08-30              2017-09-06   
118894  238-763-0612  ************4333   2017-08-31              2017-09-07   
118895  395-518-4100  ************1821   2017-08-31              2017-09-07   
118896  531-528-1017  ************7860   2017-08-31              2017-09-07   
118897  422-804-6403  ************4482   2017-08-29              2017-09-07   

       reservation_status         hotel  is_canceled  ...  \
0               Check-Out  Resort Hotel        False  ...   
1               Check-Out  Resort Hotel        False  ...   
2               Check-Out  Resort Hotel        False  ...   
3               Check-Out  Resort Hotel        False  ...   
4               Check-Out  Resort Hotel        False  ...   
...                   ...           ...          ...  ...   
118893          Check-Out    City Hotel        False  ...   
118894          Check-Out    City Hotel        False  ...   
118895          Check-Out    City Hotel        False  ...   
118896          Check-Out    City Hotel        False  ...   
118897          Check-Out    City Hotel        False  ...   

        previous_cancellations  previous_bookings_not_canceled  \
0                            0                               0   
1                            0                               0   
2                            0                               0   
3                            0                               0   
4                            0                               0   
...                        ...                             ...   
118893                       0                               0   
118894                       0                               0   
118895                       0                               0   
118896                       0                               0   
118897                       0                               0   

       reserved_room_type  assigned_room_type  booking_changes  deposit_type  \
0                       C                   C                3    No Deposit   
1                       C                   C                4    No Deposit   
2                       A                   C                0    No Deposit   
3                       A                   A                0    No Deposit   
4                       A                   A                0    No Deposit   
...                   ...                 ...              ...           ...   
118893                  A                   A                0    No Deposit   
118894                  E               

In [145]:
df.columns

Index(['Unnamed: 0', 'name', 'email', 'phone-number', 'credit_card',
       'arrival_date', 'reservation_status_date', 'reservation_status',
       'hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'num_of_weekend_nights',
       'num_of_week_nights', 'adults', 'children', 'babies', 'meal', 'country',
       'market_segment', 'distribution_channel', 'is_repeated_guest',
       'previous_cancellations', 'previous_bookings_not_canceled',
       'reserved_room_type', 'assigned_room_type', 'booking_changes',
       'deposit_type', 'days_in_waiting_list', 'customer_type',
       'average_daily_rate', 'required_car_parking_spaces'],
      dtype='object')

In [146]:

df['is_canceled'] = df['is_canceled'].apply(lambda x: x == 'True')

# Now 'is_canceled' column contains True for 'True' and False for 'False'



In [147]:
df['arrival_date'] = pd.to_datetime(df['arrival_date'])
df['reservation_status_date'] = pd.to_datetime(df['reservation_status_date'])


In [148]:
# Converting from datetime format to allow for computation

df['arrival_date'] = df['arrival_date'].dt.date
df['reservation_status_date'] = df['reservation_status_date'].dt.date

# Creating new columns 
df['time'] = df['arrival_date'] - df['reservation_status_date']

df['length_of_stay'] = df['num_of_week_nights'] + df['num_of_weekend_nights']

In [149]:
# Removing the signs from the column
df['time'] = df['time'].abs()

In [153]:
total_bookingss = df['Unnamed: 0'].count()

print(total_bookingss)

118898


In [151]:
years = df['arrival_date_year'].unique()
print(years)

[2015 2016 2017]


### Calculating individual and overall stats of booking cancellations, stay length and repeat bookings, as well as booking increase and decrease per hotel & year


In [155]:
# Calculating individual stats of booking cancellations, stay length and repeat bookings

# Total bookings
total_bookings = df.groupby(['hotel', 'arrival_date_year']).size().reset_index(name='total_bookings')

# Cancellation counts and percentages per hotel and year
cancellation_counts = df[df['reservation_status'] == 'Canceled'].groupby(['hotel', 'arrival_date_year']).size().reset_index(name='cancellation_count')

merged_df = pd.merge(total_bookings, cancellation_counts, on=['hotel', 'arrival_date_year'])

merged_df['cancellation_percentage'] = ((merged_df['cancellation_count'] / merged_df['total_bookings']) * 100).round(1)

# Average length of stay per hotel and year
avg_length_of_stay = df.groupby(['hotel', 'arrival_date_year'])['length_of_stay'].mean().round(1).reset_index(name='avg_length_of_stay')

merged_df = pd.merge(merged_df, avg_length_of_stay, on=['hotel', 'arrival_date_year'])

# Number of repeat bookings and percentages per year and hotel
repeat_bookings = df[df['is_repeated_guest'] == 1].groupby(['hotel', 'arrival_date_year']).size().reset_index(name='repeat_bookings')

merged_df = pd.merge(merged_df, repeat_bookings, on=['hotel', 'arrival_date_year'])

# Calculating repeat booking percentages
merged_df['repeat_booking_percentage'] = ((merged_df['repeat_bookings'] / merged_df['total_bookings']) * 100).round(1)

print(merged_df)


          hotel  arrival_date_year  total_bookings  cancellation_count  \
0    City Hotel               2015           13663                5864   
1    City Hotel               2016           38135               14848   
2    City Hotel               2017           27504               11448   
3  Resort Hotel               2015            8196                2054   
4  Resort Hotel               2016           18300                4798   
5  Resort Hotel               2017           13100                3938   

   cancellation_percentage  avg_length_of_stay  repeat_bookings  \
0                     42.9                 2.7              507   
1                     38.9                 3.0              915   
2                     41.6                 3.1              610   
3                     25.1                 4.5              134   
4                     26.2                 4.1              859   
5                     30.1                 4.5              781   

   repeat_b

In [154]:
# Calculating overall stats of booking cancellations, stay length and repeat bookings per hotel

percentage_total_of_cancellations = (df[df['reservation_status'] == 'Canceled'].groupby('hotel').size() / df.groupby('hotel').size() * 100).round(2)
print(percentage_total_of_cancellations)


print(df.groupby('hotel')['length_of_stay'].mean().round(2))



repeated_guest_percentage = (df[df['is_repeated_guest'] == 1].groupby('hotel').size() / df.groupby('hotel').size() * 100).round(2)
print(repeated_guest_percentage)

hotel
City Hotel      40.55
Resort Hotel    27.25
dtype: float64
hotel
City Hotel      2.98
Resort Hotel    4.34
Name: length_of_stay, dtype: float64
hotel
City Hotel      2.56
Resort Hotel    4.48
dtype: float64


In [156]:
# Calculating the booking increase and decrease per year. 

bookings_2015 = df[(df['arrival_date_year'] == 2015)]
bookings_2016 = df[(df['arrival_date_year'] == 2016)]
bookings_2017 = df[(df['arrival_date_year'] == 2017)]


# The booking increase for each hotel between 2016 and 2017
booking_increase_city = ((bookings_2016[bookings_2016['hotel'] == 'City Hotel'].shape[0] - bookings_2015[bookings_2015['hotel'] == 'City Hotel'].shape[0]) / bookings_2015[bookings_2015['hotel'] == 'City Hotel'].shape[0]) * 100

booking_increase_resort = ((bookings_2016[bookings_2016['hotel'] == 'Resort Hotel'].shape[0] - bookings_2015[bookings_2015['hotel'] == 'Resort Hotel'].shape[0]) / bookings_2015[bookings_2015['hotel'] == 'Resort Hotel'].shape[0]) * 100

# The booking decrease for each hotel between 2017 and 2016
booking_decrease_city = ((bookings_2017[bookings_2017['hotel'] == 'City Hotel'].shape[0] - bookings_2016[bookings_2016['hotel'] == 'City Hotel'].shape[0]) / bookings_2016[bookings_2016['hotel'] == 'City Hotel'].shape[0]) * 100

booking_decrease_resort = ((bookings_2017[bookings_2017['hotel'] == 'Resort Hotel'].shape[0] - bookings_2016[bookings_2016['hotel'] == 'Resort Hotel'].shape[0]) / bookings_2016[bookings_2016['hotel'] == 'Resort Hotel'].shape[0]) * 100


print(f"Booking Increase for City Hotel in 2016: {booking_increase_city:.2f}%")
print(f"Booking Increase for Resort Hotel in 2016: {booking_increase_resort:.2f}%")
print(f"Booking Decrease for City Hotel in 2017: {booking_decrease_city:.2f}%")
print(f"Booking Decrease for Resort Hotel in 2017: {booking_decrease_resort:.2f}%")


Booking Increase for City Hotel in 2016: 179.11%
Booking Increase for Resort Hotel in 2016: 123.28%
Booking Decrease for City Hotel in 2017: -27.88%
Booking Decrease for Resort Hotel in 2017: -28.42%


#### So far, we've established that:
The **City Hotel** saw a 179.11% booking increase in 2016 but suffered a −27.88% decrease in 2017, while the **Resort Hotel** saw a 123.28% booking increase in 2016 and suffered a -28.42% decrease in 2017.

**For Cancellation rates**: 40.55% of total City Hotel bookings are cancelled(42.94% in 2015, 38.97% in 2016 & 41.62% in 2017), while 27.25% of Resort Hotel bookings are cancelled(25.1% in 2015, 26.21% in 2016 & 30.11% in 2017).

**For Length of Stay**: City Hotel guests book an average of 2.98 nights(2.7 nights in 2015, 3.0 nights in 2016 & 3.1 nights in 2017), while Resort Hotel guests book an average of 4.34 nights(4.5 nights in 2015,  4.1 nights in 2016 and 4.5 nights in 2017).

**For Repeated Bookings:** Only A total of 2.56% of City Hotel bookings were repeat bookings(3.7% in 2015, 2.4% in 2016 and 2.2% in 2017), while only 4.48% of Resort Hotel bookings were repeat bookings(1.6% in 2015, 4.7% in 2016 & 6.0% in 2017).

## Research Questions
 

#### 1. Booking Cancellation Analysis:

###### Correlation doesn't equal Causation but we can find the possible factors for booking cancellations:

Average wait time and lead time for cancellers vs. non-cancellers?
What percentage of cancellers are new guests?
Which customer type dominates cancellations?
Top market segments and distribution channels for cancellations.
Top countries with the most cancellers.
Correlation between arrival season and cancellations.

#### 2. Stay Length Analysis:

###### What are the possible factors affecting length of stay:

Seasons associated with longer stays?
Top Countries with record of longer stays?
Market segments, distribution channels, and customer types accounting for longer stays?
Room types with the longer stays?
Deposit types linked to longer stays?

#### 3. Repeat Bookings Analysis:

###### What are the possible factors affecting repeat bookings:

Top arrival seasons for old vs. new guests.
Differences in lead and wait times between new and old guests.
Top countries for old and new guests.
Variances in market segments, distribution channels, and customer types for old vs. new guests.
Differences in cancellation rates and previous cancellations between old and new guests.

## Analysis

### 1. BOOKING CANCELLATIONS

#### Let's find possible reasons for the cancellation rates in both hotels.

In [157]:
# First, let's compare the average length of stay for canceled/non-canceled bookings.

avg_length_canceled = df[df['reservation_status'] == 'Canceled']['length_of_stay'].mean().round(1)
avg_length_not_canceled = df[df['reservation_status'] == 'Check-Out']['length_of_stay'].mean().round(1)


print("Average length of stay for canceled bookings:", avg_length_canceled, "days")

print("Average length of stay for non-canceled bookings:", avg_length_not_canceled, "days")



Average length of stay for canceled bookings: 3.5 days
Average length of stay for non-canceled bookings: 3.4 days


In [158]:
# Next lets's compare the average lead times for canceled/non-canceled bookings

avg_lt_canceled = df[df['reservation_status'] == 'Canceled']['lead_time'].mean().round(1)
avg_lt_not_canceled = df[df['reservation_status'] == 'Check-Out']['lead_time'].mean().round(1)

print("Average lead time for canceled bookings:", avg_lt_canceled, "days")

print("Average lead time for non-canceled bookings:", avg_lt_not_canceled, "days")


Average lead time for canceled bookings: 147.4 days
Average lead time for non-canceled bookings: 80.3 days


In [159]:
# Now, let's investigate whether seasons play a role in booking cancelations

seasons_canceled = df[df['reservation_status'] == 'Canceled']['arrival_date_month'].value_counts().nlargest(12)
seasons_not_canceled = df[df['reservation_status'] == 'Check-Out']['arrival_date_month'].value_counts().nlargest(12)

print(seasons_canceled)

print(seasons_not_canceled)


August       5142
July         4640
May          4545
June         4450
April        4435
October      4156
September    4035
March        3027
February     2466
December     2281
November     2042
January      1731
Name: arrival_date_month, dtype: int64
August       8618
July         7892
May          7102
October      6867
March        6591
April        6533
June         6393
September    6367
February     5317
November     4632
December     4365
January      4068
Name: arrival_date_month, dtype: int64


In [160]:
# Could disparity in reserved and assigned room type be a leading cause for cancelations?
disparity_in_rooms = df[df['reserved_room_type'] != df['assigned_room_type']]

# For records where reserved room is different from the room later assigned, whats's the ratio of cancelations to non-cancelations?
not_canceled_disparity_counts = disparity_in_rooms[disparity_in_rooms['reservation_status'] == 'Check-Out'].shape[0]

canceled_disparity_counts = disparity_in_rooms[disparity_in_rooms['reservation_status'] == 'Canceled'].shape[0]


print(not_canceled_disparity_counts, "bookers with disparring room type didn't cancel")

print(canceled_disparity_counts, "bookers with disparring room type canceled")


13906 bookers with disparring room type didn't cancel
588 bookers with disparring room type canceled


In [161]:
# Ascertaining if bookers from certain market segments cancel or do not cancel.

msegment_canceled = df[df['reservation_status'] == 'Canceled'].groupby('market_segment').size().sort_values(ascending=False)
msegment_not_canceled = df[df['reservation_status'] == 'Check-Out'].groupby('market_segment').size().sort_values(ascending=False)

print(msegment_canceled)

print(msegment_not_canceled)


market_segment
Online TA        20147
Groups           12023
Offline TA/TO     8047
Direct            1709
Corporate          905
Complementary       78
Aviation            41
dtype: int64
market_segment
Online TA        35664
Offline TA/TO    15882
Direct           10528
Groups            7709
Corporate         4133
Complementary      644
Aviation           185
dtype: int64


In [198]:
# Are cancelers and non-cancelers from certain distinct countries?

country_canceled = df[df['reservation_status'] == 'Canceled'].groupby('country').size().nlargest(4)
countries_not_canceled = df[df['reservation_status'] == 'Check-Out'].groupby('country').size().nlargest(4)

print(country_canceled)

print(countries_not_canceled)


country
PRT    26752
GBR     2412
ESP     2145
FRA     1902
dtype: int64
country
PRT    21071
GBR     9676
FRA     8481
ESP     6391
dtype: int64


In [199]:
# Comparing by distribution channels

distchannel_not_canceled = df[df['reservation_status'] == 'Check-Out'].groupby('distribution_channel').size().sort_values(ascending=False)
distchannel_canceled = df[df['reservation_status'] == 'Canceled'].groupby('distribution_channel').size().sort_values(ascending=False)


print(distchannel_not_canceled )


print(distchannel_canceled)


distribution_channel
TA/TO        57611
Direct       11940
Corporate     5037
GDS            156
Undefined        1
dtype: int64
distribution_channel
TA/TO        39235
Direct        2317
Corporate     1364
GDS             34
dtype: int64


In [164]:
# Comparing groups by customer type and year

type_customer_canceled = df[df['reservation_status'] == 'Canceled'].groupby(['customer_type','arrival_date_year']).size()
type_customer_not_canceled = df[df['reservation_status'] == 'Check-Out'].groupby(['customer_type','arrival_date_year']).size()


print(type_customer_canceled)

print(type_customer_not_canceled)

customer_type    arrival_date_year
Contract         2015                  1047
                 2016                   115
                 2017                    74
Group            2015                    32
                 2016                    10
                 2017                    12
Transient        2015                  3991
                 2016                 17182
                 2017                 14334
Transient-Party  2015                  2848
                 2016                  2339
                 2017                   966
dtype: int64
customer_type    arrival_date_year
Contract         2015                  1597
                 2016                   632
                 2017                   585
Group            2015                   158
                 2016                   152
                 2017                   202
Transient        2015                  7170
                 2016                 26438
                 2017                

In [165]:
# Does booking changes affect booking cancelations or not?

canceled_reservations = df[df['reservation_status'] == 'Canceled']
notcanceled_reservations = df[df['reservation_status'] == 'Check-Out']


avg_booking_changes_canceled = canceled_reservations['booking_changes'].sum().round(1)
avg_booking_changes_not_canceled = notcanceled_reservations['booking_changes'].sum().round(1)

print("Average booking changes for canceled reservations:", avg_booking_changes_canceled)
print("Average booking changes for non-canceled reservations:", avg_booking_changes_not_canceled)

Average booking changes for canceled reservations: 4061
Average booking changes for non-canceled reservations: 21957


In [166]:
# Does wait times affect booking cancelations?

canceled = df[df['reservation_status'] == 'Canceled']
not_canceled = df[df['reservation_status'] == 'Check-Out']


waiting_list_canceled = canceled['days_in_waiting_list'].mean().round(1)
waiting_list_not_canceled = not_canceled['days_in_waiting_list'].mean().round(1)

print("Average days in waiting list for those who canceled:", waiting_list_canceled, "days.")
print("Average days in waiting list for those who canceled:", waiting_list_not_canceled, "days.")

Average days in waiting list for those who canceled: 3.7 days.
Average days in waiting list for those who canceled: 1.6 days.


In [200]:
# Are cancelers old or new guests?

canceled_old_guests = df[(df['reservation_status'] == 'Canceled') & (df['is_repeated_guest'] == 1)].value_counts().shape[0]
canceled_new_guests = df[(df['reservation_status'] == 'Canceled') & (df['is_repeated_guest'] == 0)].value_counts().shape[0]
not_canceled_old_guests = df[(df['reservation_status'] == 'Check-Out') & (df['is_repeated_guest'] == 1)].value_counts().shape[0]
not_canceled_new_guests = df[(df['reservation_status'] == 'Check-Out') & (df['is_repeated_guest'] == 0)].value_counts().shape[0]


print(canceled_old_guests, "bookers who canceled have previously booked ")

print(canceled_new_guests, "bookers who canceled have not previously booked ")

print(not_canceled_old_guests, "bookers who didn't cancel have previously booked")


print(not_canceled_new_guests, "bookers who didn't cancel have not previously booked")


507 bookers who canceled have previously booked 
42443 bookers who canceled have not previously booked 
3254 bookers who didn't cancel have previously booked
71491 bookers who didn't cancel have not previously booked


### FINDINGS A : GUESTS CANCELLATION INSIGHTS

- From the result above, we see that the length of stay booked by guests may most likely not play a role in booking cancellations. That is, there's no indication that guests cancel because of proposed length of stay as there is hardly a difference between that of canceled and non-canceled bookings.
- There is a 67.1 days difference between lead times for canceled and non-canceled bookings. That's enough difference to suggest that lead times may play a role in booking cancelations.
- The summer months, for both categories record the highest bookings, while winter months hold the least.
- Although a wide margin, the result shows that disparity in room types is likely NOT the leading cause of booking cancelations.
- For both categories, Online TA emerges top and Aviation least. 
  Therefore, market segments likely do not play a role in cancelations
- There are no differences in the 4 top countries(PRT, GBR, ESP & FRA), guests from these 4 countries are both the main cancellers and non-cancellers.
- In the two groups, the top and least distribution channels are similar.
- Given the data, Transient customer type both canceled the most and didn't cancel the most, as well.
- Booking changes is not likely a leading cause of cancelations as there are more than 100% higher booking changes for non-cancelers than cancelers.
- There's up to 2 days in waiting list difference for those who canceled. It is possible for guests to have opted out for lack of impatience or they found other alternatives that doesn't require waiting.
- Both cancelers and non-cancelers are new and old guests. In both categories new guests are much more.

### 2. LENGTH OF STAY 

#### What factors influence guest's length of stay?

In [201]:
# Are there significant differnces in booked length of stay between overall bookers, cancelers and non-cancelers?
avg_length_of_stay_all = df.groupby(['hotel','arrival_date_year'])['length_of_stay'].mean().round(1)

#Filtered
noncancelers_filtered = df[df['reservation_status'] == 'Check-Out'].groupby(['hotel','arrival_date_year'])
avg_length_of_stay_noncancelers = noncancelers_filtered['length_of_stay'].mean().round(1)

cancelers_filtered = df[df['reservation_status'] == 'Canceled'].groupby(['hotel','arrival_date_year'])
avg_length_of_stay_cancelers = cancelers_filtered['length_of_stay'].mean().round(1)

print(avg_length_of_stay_all)
print(avg_length_of_stay_noncancelers)
print(avg_length_of_stay_cancelers)

hotel         arrival_date_year
City Hotel    2015                 2.7
              2016                 3.0
              2017                 3.1
Resort Hotel  2015                 4.5
              2016                 4.1
              2017                 4.5
Name: length_of_stay, dtype: float64
hotel         arrival_date_year
City Hotel    2015                 2.8
              2016                 2.9
              2017                 3.0
Resort Hotel  2015                 4.4
              2016                 4.0
              2017                 4.3
Name: length_of_stay, dtype: float64
hotel         arrival_date_year
City Hotel    2015                 2.7
              2016                 3.0
              2017                 3.2
Resort Hotel  2015                 4.7
              2016                 4.6
              2017                 5.2
Name: length_of_stay, dtype: float64


In [202]:
# Do seasons affect guests' length of stay?
resort_season_length = df[df['hotel'] == 'Resort Hotel'].groupby(['arrival_date_month'])['length_of_stay'].mean().round(1).nlargest(12)
city_season_length = df[df['hotel'] == 'City Hotel'].groupby(['arrival_date_month'])['length_of_stay'].mean().round(1).nlargest(12)
resort_length = resort_season_length.rename('Average Length of Stay for Resort Hotel')
city_length = city_season_length.rename('Average Length of Stay for City Hotel')

print(resort_length)
print(city_length)

arrival_date_month
June         5.4
July         5.3
August       5.2
September    5.1
May          4.3
March        4.2
April        4.1
October      4.0
November     3.6
December     3.3
February     3.1
January      2.9
Name: Average Length of Stay for Resort Hotel, dtype: float64
arrival_date_month
August       3.2
December     3.2
July         3.1
March        3.1
April        3.0
February     3.0
January      3.0
November     3.0
June         2.9
May          2.8
September    2.8
October      2.7
Name: Average Length of Stay for City Hotel, dtype: float64


In [170]:
# Now let's see if there is any correlation between the lead time and length of stay
corr_length_lead_time = df['length_of_stay'].corr(df['lead_time']).round(1)

# Is there any correlation between the age categories of guests and length of stay

corr_length_adults = df['length_of_stay'].corr(df['adults']).round(1)

corr_length_children = df['length_of_stay'].corr(df['children']).round(1)

corr_length_babies = df['length_of_stay'].corr(df['babies']).round(1)

print("The correlation between lead time and length of stay is:", corr_length_lead_time)
print("The correlation between adults and length of stay is:",corr_length_adults)
print("The correlation between children and length of stay is:",corr_length_children)
print("The correlation between babies and length of stay is:",corr_length_babies)

The correlation between lead time and length of stay is: 0.2
The correlation between adults and length of stay is: 0.1
The correlation between children and length of stay is: 0.1
The correlation between babies and length of stay is: 0.0


In [215]:
# Do guests from certain countries stay longer at the Resort hotel?

country_length_resort_large = df[df['hotel'] == 'Resort Hotel'].groupby('country')['length_of_stay'].mean().nlargest(5)
country_length_resort_small = df[df['hotel'] == 'Resort Hotel'].groupby('country')['length_of_stay'].mean().nsmallest(5)
length_large_resort = country_length_resort_large.rename('Longest Average Length of stay by country for Resort Hotel')
length_small_resort = country_length_resort_small.rename('Shortest Average Length of stay by country for Resort Hotel')

print(length_large_resort)
print(length_small_resort)

# Do guests from certain countries stay longer at the City Hotel?

country_length_city_large = df[df['hotel'] == 'City Hotel'].groupby('country')['length_of_stay'].mean().round(1).nlargest(5)
country_length_city_small = df[df['hotel'] == 'City Hotel'].groupby('country')['length_of_stay'].mean().round(1).nsmallest(5)
length_large_city = country_length_city_large.rename('Longest Average Length of stay by country for City Hotel')
length_small_city = country_length_city_small.rename('Shortest Average Length of stay by country for City Hotel')

print(length_large_city)
print(length_small_city)



country
SEN    10.0
AZE     9.0
TGO     9.0
MKD     8.0
ARM     7.0
Name: Longest Average Length of stay by country for Resort Hotel, dtype: float64
country
QAT    0.0
BIH    1.0
COM    1.0
DJI    1.0
GGY    1.0
Name: Shortest Average Length of stay by country for Resort Hotel, dtype: float64
country
FRO    12.0
SEN     8.7
AGO     8.5
CPV     7.2
GNB     7.1
Name: Longest Average Length of stay by country for City Hotel, dtype: float64
country
ASM    1.0
BFA    1.0
COM    1.0
HND    1.0
MLI    1.0
Name: Shortest Average Length of stay by country for City Hotel, dtype: float64


In [213]:
# Do guests belonging to ceratin market segments stay for longer nights?

market_length_city = df[df['hotel'] == 'City Hotel'].groupby(['market_segment'])['length_of_stay'].mean().round(1).sort_values(ascending=False)
market_length_resort = df[df['hotel'] == 'Resort Hotel'].groupby(['market_segment'])['length_of_stay'].mean().round(1).sort_values(ascending=False)
length_market_city = market_length_city.rename('Average Length of stay for City hotel by market segment')
length_market_resort = market_length_resort.rename('Average Length of stay for Resort hotel by market segment')

print(length_market_city)
print(length_market_resort)

market_segment
Aviation         3.6
Online TA        3.3
Direct           2.9
Offline TA/TO    2.9
Groups           2.5
Corporate        1.9
Complementary    1.5
Name: Average Length of stay for City hotel by market segment, dtype: float64
market_segment
Offline TA/TO    6.2
Online TA        4.2
Groups           4.1
Direct           3.5
Corporate        2.4
Complementary    2.1
Name: Average Length of stay for Resort hotel by market segment, dtype: float64


In [212]:
# Comparing length of stay by distribution channels

distchannel_length_resort = df[df['hotel'] == 'Resort Hotel'].groupby(['distribution_channel'])['length_of_stay'].mean().round(1).sort_values(ascending=False)
distchannel_length_city = df[df['hotel'] == 'City Hotel'].groupby(['distribution_channel'])['length_of_stay'].mean().round(1).sort_values(ascending=False)
length_distchannel_resort = distchannel_length_resort.rename('Average Length of stay for Resort hotel by distribution channel')
length_distchannel_city = distchannel_length_city.rename('Average Length of stay for City hotel by distribution channel')

print(length_distchannel_resort)
print(length_distchannel_city)



distribution_channel
Undefined    5.0
TA/TO        4.7
Direct       3.6
Corporate    2.9
Name: Average Length of stay for Resort hotel by distribution channel, dtype: float64
distribution_channel
TA/TO        3.1
Direct       2.8
Corporate    2.0
GDS          2.0
Name: Average Length of stay for City hotel by distribution channel, dtype: float64


In [207]:
# Could there be a correlation between length of stay and guests' previous cancelations?

corr_prevcanc_city = df[df['hotel'] == 'City Hotel']['length_of_stay'].corr(df['previous_cancellations']).round(1)
corr_prevcanc_resort = df[df['hotel'] == 'Resort Hotel']['length_of_stay'].corr(df['previous_cancellations']).round(1)

print("The correlation between stay length and previous cancelation in City hotel is ", corr_prevcanc_city)
print("The correlation between stay length and previous cancelation in Resort hotel is",corr_prevcanc_resort)

The correlation between stay length and previous cancelation in City hotel is  -0.1
The correlation between stay length and previous cancelation in Resort hotel is -0.0


In [208]:
# Checking for correlation between length of stay and previous bookings not canceled, booking changes, days in waiting list

corr_length_prevnotcanc = df['length_of_stay'].corr(df['previous_bookings_not_canceled']).round(1)
corr_length_booking_changes = df['length_of_stay'].corr(df['booking_changes']).round(1)
corr_length_waiting_list = df['length_of_stay'].corr(df['days_in_waiting_list']).round(1)


print("The correlation between stay length and previous booking not canceled is ", corr_length_prevnotcanc)

print("The correlation between stay length and booking changes is ", corr_length_booking_changes)


print("The correlation between stay length and days in waiting list is ", corr_length_waiting_list)

The correlation between stay length and previous booking not canceled is  -0.1
The correlation between stay length and booking changes is  0.1
The correlation between stay length and days in waiting list is  -0.0


In [209]:
# What is the average stay length per customer type for each hotel

cust_length_resort = df[df['hotel'] == 'Resort Hotel'].groupby(['customer_type'])['length_of_stay'].mean().round(1).sort_values(ascending=False)
cust_length_city = df[df['hotel'] == 'City Hotel'].groupby(['customer_type'])['length_of_stay'].mean().round(1).sort_values(ascending=False)
length_custtype_resort = cust_length_resort.rename('Length of stay for Resort hotel by Customer type')
length_custtype_city = cust_length_city.rename('Length of stay for City hotel by Customer type')

print(length_custtype_resort)
print(length_custtype_city)



customer_type
Contract           8.6
Transient          4.2
Transient-Party    3.9
Group              3.2
Name: Length of stay for Resort hotel by Customer type, dtype: float64
customer_type
Transient          3.1
Contract           2.8
Transient-Party    2.7
Group              2.6
Name: Length of stay for City hotel by Customer type, dtype: float64


In [211]:
# Let's see the average length of stay by rooms reserved.
city_filtered = df[df['hotel'] == 'City Hotel']
resort_filtered = df[df['hotel'] == 'Resort Hotel']


resroom_length_city = city_filtered.groupby(['reserved_room_type'])['length_of_stay'].mean().round(1).sort_values(ascending=False)
resroom_length_resort = resort_filtered.groupby(['reserved_room_type'])['length_of_stay'].mean().round(1).sort_values(ascending=False)
length_resroom_city = resroom_length_city.rename('Average Length of stay by Reserved room in City hotel')
length_resroom_resort = resroom_length_resort.rename('Average Length of stay by Reserved room in Resort hotel')


print(length_resroom_city)
print(length_resroom_resort)

reserved_room_type
D    3.5
E    3.5
B    3.4
G    3.4
F    3.3
A    2.9
C    2.1
P    0.0
Name: Average Length of stay by Reserved room in City hotel, dtype: float64
reserved_room_type
D    5.2
E    5.1
C    4.7
G    4.6
F    4.4
H    4.0
A    3.9
B    1.3
L    1.2
Name: Average Length of stay by Reserved room in Resort hotel, dtype: float64


In [217]:
# Lastly, let us see stay length per deposit type.
deptype_length_city = df[df['hotel'] == 'City Hotel'].groupby(['deposit_type'])['length_of_stay'].mean().round(1).sort_values(ascending=False)
deptype_length_resort = df[df['hotel'] == 'Resort Hotel'].groupby(['deposit_type'])['length_of_stay'].mean().round(1).sort_values(ascending=False)
length_deptype_city = deptype_length_city.rename('Average Length of stay by Deposit type in City hotel')
length_deptype_resort = deptype_length_resort.rename('Average Length of stay by Deposit type in Resort hotel')


print(length_deptype_city)


print(length_deptype_resort)


deposit_type
Refundable    4.5
No Deposit    3.1
Non Refund    2.6
Name: Average Length of stay by Deposit type in City hotel, dtype: float64
deposit_type
No Deposit    4.4
Non Refund    3.9
Refundable    3.7
Name: Average Length of stay by Deposit type in Resort hotel, dtype: float64


### FINDINGS B: FACTORS INFLUENCING GUESTS LENGTH OF STAY

- There are no significant differences for each group- Overall, canceled guests and non-canceled guests.

- There is no or weak correlation between the length of stay and lead time, if guests are adults, children or babies.

- For the Resort, guests typically stay longer(5-4 nights) during the summer, fall and spring seasons, while winter season records the shortest stays(3-2 nights).
  For the City hotel, the hot summer and the festive december takes the lead for the longer stays(3.2 nights), while Fall seasons holds the shorter stays with 2 nights.
  
- Guests who stay at the Resort Hotel the longest(an average of 7-10 nights) come from Senegal, Azerbaijan, Togo, North Macedonia and Armenia; 
  while the ones who stay for less than a night to 1 night come from Qatar, Bosnia & Herzegovina, Comoros, Djibouti and Guernsey
- Guests who stay at the City Hotel the longest(average of 7-12 nights) come from Faroe Islands, Senegal , Angola, Cape Verde and Guinea-Bissau; 
  while the ones who stay for only a night to 1 night come from American Samoa, Burkina Faso, Comoros, Honduras and Mali.
  
- For the City hotel, guests falling into the Aviation and Online TA market segments, booked the longest stays with an average of 3 nights and guests from the Corporate & Complimentary market segments, booked the shortest stays, averaging a single night.
  While Resort hotel guests, who fall into the Offline TA/TO , Online TA market as well as Group market segments, booked the longest stays with an average of 4-6 nights. Guests from the Corporate & Complimentary market segments also, booked the shortest stays, averaging 2 nights.
  
- Resort hotel guests who booked through Undefined and TA/TO distribution channels, account for guests with the longest stays with an average of 4-5 nights and those who booked through Corporate distribution channel, booked the shortest stays, averaging 2 nights. 
  While City hotel guests, who booked via TA/TO and Direct distribution channels, booked the longest stays with an average of 2.8 - 3 nights. Guests who booked via the Corporate & GDS distribution channels, booked the shortest stays, averaging 2 nights.
  
- A negative correlation is indicated for both Resort and City hotel, meaning there's no correlation between previous cancelation and length of stay.

- It is safe to say there is no positive correlation between length of stay and previous bookings not canceled, booking changes and days in waiting list.

- Contract customers stay longer at the Resort hotel, an average of 8 nights, and Group customers stay have the shortest with only an average of 3 nights.
  Transient customers stay the longest at the City hotel with an average of 3 nights and Group customers stay the least also with an average of 2 nights.
  
- In the City hotel, guests who booked Triple(D) and Single(E) rooms stayed the longest for an average of 3.5 nights, guests in the Presidential Suite (P) didn't stay for up to a night.
  In the Resort hotel, guests who booked Triple(D) and Single(E) rooms also stayed the longest for an average of 5 nights, whereas Superior Room (B) and Deluxe Room (L) occupants stayed the shortest for only an average of a single night.
  
- In the City hotel, guests who utilize refundable deposits, tend to book longer stays of an average of 4 nights, guests who use non-refundable book shorter for 2 nights.
  In Resort hotel, guests without deposit book for longest( an average of 4 nights) while those with refundable deposits book shorter stays(3 nights).

### 3. TRENDS IN CUSTOMER RETENTION & LOYALTY

In [218]:
# Comparing seasons for repeat bookings
szn_repeated_resort = df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] == True)].groupby('arrival_date_month').size().nlargest(12)
szn_repeated_city = df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] == True)].groupby('arrival_date_month').size().nlargest(12)

szn_not_repeated_resort = df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] == False)].groupby('arrival_date_month').size().nlargest(12)
szn_not_repeated_city = df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] == False)].groupby('arrival_date_month').size().nlargest(12)

final_szn_repeated_resort = szn_repeated_resort.rename("Repeated guests by seasons in Resort hotel")
final_szn_not_repeated_city = szn_not_repeated_city.rename("New guests by seasons in City hotel")
final_szn_repeated_city = szn_repeated_city.rename("Repeated guests by seasons in City hotel")
final_szn_not_repeated_resort = szn_not_repeated_resort.rename("New guests by seasons in Resort hotel")


print(final_szn_repeated_resort)
print(final_szn_not_repeated_city)
print(final_szn_repeated_city)
print(final_szn_not_repeated_resort)

arrival_date_month
February     265
March        242
January      219
April        178
May          163
December     146
November     129
June         127
August        85
July          85
October       83
September     52
Name: Repeated guests by seasons in Resort hotel, dtype: int64
arrival_date_month
August       8842
May          8029
July         7872
June         7739
April        7371
October      7361
September    7251
March        6275
February     4819
November     4190
December     3963
January      3558
Name: New guests by seasons in City hotel, dtype: int64
arrival_date_month
October      230
July         216
May          203
March        183
January      178
December     166
November     164
June         155
September    149
February     146
August       137
April        105
Name: Repeated guests by seasons in City hotel, dtype: int64
arrival_date_month
August       4788
July         4455
October      3421
April        3391
May          3384
March        3039
September   

In [219]:
# Does lead time affect guests' retention?

lead_time_resort_retend = df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] == True)]['lead_time'].mean().round(1)
lead_time_city_retend = df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] == True)]['lead_time'].mean().round(1)
lead_time_resort_unretend = df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] == False)]['lead_time'].mean().round(1)
lead_time_city_unretend = df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] == False)]['lead_time'].mean().round(1)



print("The average lead time for Resort hotel old guests is", lead_time_resort_retend)
,
print("The average lead time for City hotel old guests is", lead_time_city_retend)

print("The average lead time for Resort hotel new guests is", lead_time_resort_unretend)

print("The average lead time for City hotel new guests is", lead_time_city_unretend)

The average lead time for Resort hotel old guests is 24.9
The average lead time for City hotel old guests is 35.9
The average lead time for Resort hotel new guests is 96.7
The average lead time for City hotel new guests is 111.7


In [221]:
# Let's see the distribution of old guests as it relates to customer age category.

adults_rentention = df[df['is_repeated_guest'] == True].groupby('hotel')['adults'].sum()
children_rentention = df[df['is_repeated_guest'] == True].groupby('hotel')['children'].sum()
babies_rentention = df[df['is_repeated_guest'] == True].groupby('hotel')['babies'].sum()


print(adults_rentention)

print(children_rentention)

print(babies_rentention)

hotel
City Hotel      2693
Resort Hotel    2594
Name: adults, dtype: int64
hotel
City Hotel      51
Resort Hotel    70
Name: children, dtype: int64
hotel
City Hotel      8
Resort Hotel    4
Name: babies, dtype: int64


In [183]:
# Does country play a factor in repeat booking?

old_guest_countries = df[df['is_repeated_guest'] == True].groupby('country').size().nlargest(2)
new_guest_countries = df[df['is_repeated_guest'] == False].groupby('country').size().nlargest(2)
print(old_guest_countries)
print(new_guest_countries)

country
PRT    3250
GBR     107
dtype: int64
country
PRT    45336
GBR    12022
dtype: int64


In [184]:
# Comparing by market segments

print(df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] ==True)].groupby('market_segment').size().sort_values(ascending=False))
print(df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] ==True)].groupby('market_segment').size().sort_values(ascending=False))

print(df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] ==False)].groupby('market_segment').size().sort_values(ascending=False))
print(df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] ==False)].groupby('market_segment').size().sort_values(ascending=False))



market_segment
Direct           603
Corporate        560
Online TA        364
Offline TA/TO    160
Groups            69
Complementary     18
dtype: int64
market_segment
Corporate        913
Online TA        232
Offline TA/TO    219
Complementary    212
Groups           201
Direct           191
Aviation          64
dtype: int64
market_segment
Online TA        17291
Offline TA/TO     7267
Groups            5763
Direct            5754
Corporate         1566
Complementary      181
dtype: int64
market_segment
Online TA        38515
Offline TA/TO    16514
Groups           13773
Direct            5900
Corporate         2072
Complementary      323
Aviation           173
dtype: int64


In [185]:
# Comparing by distribution channels for Resort Hotel with repeated guests
resort_repeated_channels = df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] == True)].groupby('distribution_channel').size().sort_values(ascending=False)
print("Resort Hotel with repeated guests:")
print(resort_repeated_channels)

# Comparing by distribution channels for City Hotel with repeated guests
city_repeated_channels = df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] == True)].groupby('distribution_channel').size().sort_values(ascending=False)
print("\nCity Hotel with repeated guests:")
print(city_repeated_channels)

# Comparing by distribution channels for Resort Hotel with non-repeated guests
resort_nonrepeated_channels = df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] == False)].groupby('distribution_channel').size().sort_values(ascending=False)
print("\nResort Hotel with non-repeated guests:")
print(resort_nonrepeated_channels)

# Comparing by distribution channels for City Hotel with non-repeated guests
city_nonrepeated_channels = df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] == False)].groupby('distribution_channel').size().sort_values(ascending=False)
print("\nCity Hotel with non-repeated guests:")
print(city_nonrepeated_channels)


Resort Hotel with repeated guests:
distribution_channel
TA/TO        639
Direct       571
Corporate    564
dtype: int64

City Hotel with repeated guests:
distribution_channel
Corporate    1007
TA/TO         659
Direct        363
GDS             3
dtype: int64

Resort Hotel with non-repeated guests:
distribution_channel
TA/TO        28161
Direct        7134
Corporate     2526
Undefined        1
dtype: int64

City Hotel with non-repeated guests:
distribution_channel
TA/TO        68271
Direct        6415
Corporate     2394
GDS            190
dtype: int64


In [232]:
# Resort Hotel: repeated guests with previous cancellations
resort_repeated_prevcancellations = df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] == True)]['previous_cancellations'].sum()
print("Resort Hotel repeated guests with previous cancellations:", resort_repeated_prevcancellations)

# City Hotel: repeated guests with previous cancellations
city_repeated_prevcancellations = df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] == True)]['previous_cancellations'].sum()
print("City Hotel repeated guests with previous cancellations:", city_repeated_prevcancellations)

# Resort Hotel: non-repeated guests with previous cancellations
resort_nonrepeated_prevcancellations = df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] == False)]['previous_cancellations'].sum()
print("Resort Hotel non-repeated guests with previous cancellations:", resort_nonrepeated_prevcancellations)

# City Hotel: non-repeated guests with previous cancellations
city_nonrepeated_prevcancellations = df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] == False)]['previous_cancellations'].sum()
print("City Hotel non-repeated guests with previous cancellations:", city_nonrepeated_prevcancellations)


Resort Hotel repeated guests with previous cancellations: 231
City Hotel repeated guests with previous cancellations: 1559
Resort Hotel non-repeated guests with previous cancellations: 3804
City Hotel non-repeated guests with previous cancellations: 4767


In [233]:
total_cancellations_repeated = df[df['is_repeated_guest'] == True]['previous_cancellations'].sum()
total_guests_repeated = len(df[df['is_repeated_guest'] == True])

total_cancellations_non_repeated = df[df['is_repeated_guest'] == False]['previous_cancellations'].sum()
total_guests_non_repeated = len(df[df['is_repeated_guest'] == False])

cancellation_rate_repeated = total_cancellations_repeated / total_guests_repeated
cancellation_rate_non_repeated = total_cancellations_non_repeated / total_guests_non_repeated

print(f'Previous Cancellation rate for repeated guests: {cancellation_rate_repeated:.2%}')
print(f'Previous Cancellation rate for non-repeated guests: {cancellation_rate_non_repeated:.2%}')


Previous Cancellation rate for repeated guests: 47.03%
Previous Cancellation rate for non-repeated guests: 7.45%


In [234]:
# The total booking changes for repeated guests
total_booking_changes_repeat = df[df['is_repeated_guest'] == True]['booking_changes'].sum()

# Resort Hotel: repeated guests' booking changes
resort_repeated_booking_changes = df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] == True)]['booking_changes'].sum()
percentage_resort_repeated_booking_changes = (resort_repeated_booking_changes / total_booking_changes_repeat) * 100
print("Percentage of total booking changes for Old Resort Hotel guests:", round(percentage_resort_repeated_booking_changes, 2), "%")

# City Hotel: repeated guests' booking changes
city_repeated_booking_changes = df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] == True)]['booking_changes'].sum()
percentage_city_repeated_booking_changes = (city_repeated_booking_changes / total_booking_changes_repeat) * 100
print("Percentage of total booking changes for Old City Hotel guests:", round(percentage_city_repeated_booking_changes, 2), "%")

# The total booking changes for new guests
total_booking_changes_nonrepeat = df[df['is_repeated_guest'] == False]['booking_changes'].sum()

# Resort Hotel: new guests' booking changes
resort_nonrepeated_booking_changes = df[(df['hotel'] == 'Resort Hotel') & (df['is_repeated_guest'] == False)]['booking_changes'].sum()
percentage_resort_nonrepeated_booking_changes = (resort_nonrepeated_booking_changes / total_booking_changes_nonrepeat) * 100
print("Percentage of total booking changes for new Resort Hotel guests:", round(percentage_resort_nonrepeated_booking_changes, 2), "%")

# City Hotel: new guests' booking changes
city_nonrepeated_booking_changes = df[(df['hotel'] == 'City Hotel') & (df['is_repeated_guest'] == False)]['booking_changes'].sum()
percentage_city_nonrepeated_booking_changes = (city_nonrepeated_booking_changes / total_booking_changes_nonrepeat) * 100
print("Percentage of total booking changes for new City Hotel guests:", round(percentage_city_nonrepeated_booking_changes, 2), "%")


Percentage of total booking changes for Old Resort Hotel guests: 51.39 %
Percentage of total booking changes for Old City Hotel guests: 48.61 %
Percentage of total booking changes for new Resort Hotel guests: 43.16 %
Percentage of total booking changes for new City Hotel guests: 56.84 %


In [231]:
total_changes_repeated = df[df['is_repeated_guest'] == True]['booking_changes'].sum()
total_guests_repeated_changes = len(df[df['is_repeated_guest'] == True])

total_changes_non_repeated = df[df['is_repeated_guest'] == False]['booking_changes'].sum()
total_guests_non_repeated_changes = len(df[df['is_repeated_guest'] == False])

changes_rate_repeated = total_changes_repeated / total_guests_repeated_changes
changes_rate_non_repeated = total_changes_non_repeated / total_guests_non_repeated_changes

print(f'Booking Changes rate for repeated guests: {changes_rate_repeated:.2%}')
print(f'Booking Changes rate for non-repeated guests: {changes_rate_non_repeated:.2%}')


Booking Changes rate for repeated guests: 26.48%
Booking Changes rate for non-repeated guests: 21.97%


In [226]:
# Repeated guests and customer type per hotel
repeated_guests_customer_type = df[df['is_repeated_guest'] == True].groupby(['customer_type', 'hotel']).size().sort_values(ascending=False)
print("Repeated Guests Grouped by Customer Type and Hotel:")
print(repeated_guests_customer_type)

# Non-repeated guests and customer per hotel
nonrepeated_guests_customer_type = df[df['is_repeated_guest'] == False].groupby(['customer_type', 'hotel']).size().sort_values(ascending=False)
print("\nNon-Repeated Guests Grouped by Customer Type and Hotel:")
print(nonrepeated_guests_customer_type)


Repeated Guests Grouped by Customer Type and Hotel:
customer_type    hotel       
Transient        City Hotel      1618
                 Resort Hotel    1452
Transient-Party  City Hotel       352
                 Resort Hotel     190
Group            Resort Hotel     113
                 City Hotel        48
Contract         Resort Hotel      19
                 City Hotel        14
dtype: int64

Non-Repeated Guests Grouped by Customer Type and Hotel:
customer_type    hotel       
Transient        City Hotel      57764
                 Resort Hotel    28340
Transient-Party  City Hotel      16976
                 Resort Hotel     7560
Contract         City Hotel       2286
                 Resort Hotel     1757
Group            City Hotel        244
                 Resort Hotel      165
dtype: int64


In [225]:

# Repeated guests and deposit type
repeated_guests_deposit = df[df['is_repeated_guest'] == True].groupby('deposit_type').size().sort_values(ascending=False)
print("Repeated Guests Grouped by Deposit Type:")
print(repeated_guests_deposit)

# Non-repeated guests and deposit type
nonrepeated_guests_deposit = df[df['is_repeated_guest'] == False].groupby('deposit_type').size().sort_values(ascending=False)
print("\nNon-Repeated Guests Grouped by Deposit Type:")
print(nonrepeated_guests_deposit)


Repeated Guests Grouped by Deposit Type:
deposit_type
No Deposit    3738
Non Refund      64
Refundable       4
dtype: int64

Non-Repeated Guests Grouped by Deposit Type:
deposit_type
No Deposit    100425
Non Refund     14509
Refundable       158
dtype: int64


### FINDINGS C: GUESTS RENTENTION PATTERNS

- Top arrival seasons for old guests in City hotel is in Fall , the spring season records the least. Top arrival seasons for new guests in City hotel is in Summer and late Spring , winter season records the least.
  Top arrival seasons for old guests in Resort hotel is in Late winter and Early Spring , early fall records the least while new guests arrive mostly by summer and least by winter.
 
- The lead time for old and new guests in both hotels is significantly differing. 

- The vast majority in each of these age categories are new guests.

- Both old and new guests come mainly from Portugal and the United Kingdom.

- For old guests in Resort hotel, Directis top market segment and Complimentary is least. In City hotel, old guests belong mostly to the Corporate market segment and aviation the least.
  For new guests in Resort hotel, Online TA tops the market segment and Complementary is least. In City hotel, Online TA/TO tops the segment and Avation is least.
  
- Old guests in Resort hotel mainly booked through TA/TO, while those in City hotel booked via Corporate.
  New guests in Resort hotel and City Hotel mainly booked via TA/TO. 

- Old guests have significantly higher cancelation rates than new guests.
  
- There are no significant differences between the booking changes for each categories in both hotels.

- Both Old and new guests are mostly Transient customers, in both hotels.

- The majority of both Old and new guests had no deposits


In [222]:
df.to_csv('hotelnew_data.csv')

## Conclusion & Recommendations

### Recommended strategies to reduce booking cancellations➖

Given that long lead and wait times likely play a role in reason for cancellations, I recommend that the following measures be employed:- 

- The IT and Systems Team should **streamline** the online booking process like infusing real-time availability and automated confirmation to minimize waiting times for guests. 
- Communications Team should regularly engage with guests through personalized welcome messages, strategic newsletters or updates about special offers and events, to keep guests motivated about their upcoming stay and well informed about hotel amenities, services and experience, as well as local attractions.
- Optimize the Travel-Agencies/Tour-Operators channels, through the provision of targeted incentives, deals as well as clear cancellation policies. A feedback system or Surveys should be introduced to better understand the reasons behind cancellations.
- Encourage guests to book by offering exclusive deals, promotions and packages during the Summer and Spring seasons.
- Dynamic pricing strategies that adjust rates based on the demand and other factors should be explored, as offering competitive pricing during peak cancellation periods may attract more bookings.

### Recommended Strategies to increase length of stay

I’ve observed some factors that may be responsible for length of stay, based on this, I recommend the following:- 

- Capitalizing on stays peak period-summer and December for City Hotel and Summer, Fall and Spring for Resort hotel, offer seasonal packages and promotions to attract guests during these peak periods. Focus marketing efforts on attracting guests from Senegal, Azerbaijan, Togo, North Macedonia, and Armenia for Resort hotel and Faroe Islands, Senegal, Angola, Cape Verde, and Guinea-Bissau for City Hotel, given that they tend to stay the longest; through tailored marketing campaigns, partnerships, or promotions targeting these specific nationalities to increase their engagement.
- Tailor marketing strategies for the top market segments—Offline TA/TO, Online TA, and Group segments for Resort Hotel, and Aviation and Online TA segments for City Hotel, introduce extended stay packages and benefits, loyalty programs or exclusive perks for guests in these segments to encourage longer bookings.
- Since guests booking through TA/TO channels for Resort Hotel and Direct and TA/TO for City hotel tend to stay longer, we should focus on increasing visibility, special deals, discounts and promotions through these channels. We should provide incentives for bookings through these channels. And as for channels with lower stays, like Corporate & GDS distribution, explore strategies to incentivize longer stays.
- Engage with Contract & Transient customers by offering personalized services or exclusive amenities for their longer stays. For Group customers, consider creating packages that encourage them to extend their stays with group-friendly activities and benefits, emphasizing the City & Resort Hotel's amenities and attractions for a longer and more enjoyable stay.
- Promote Triple (D) and Single (E) rooms for both Hotels, as ideal choices for guests seeking longer stays, emphasizing their comfort and features. Also consider special rates or packages for these room type. Rooms with shorter stays, Superior Room (B) and Deluxe Room (L) for the Resort Hotel as well as the Presidential Suite (P) for the City Hotel should be paired with exclusive experiences or perks.
- Collect feedback from guests who both stayed for shorter or longer durations and use it to identify areas for improvement. Address concerns and make enhancements to the guest experience, amenities, or services to sustain and encourage longer stays.
- Collaborate with local businesses or attractions to create attractive packages that encourage longer stays. Offer bundled services, such as spa treatments, guided tours, or dining experiences, to entice guests to extend their visit.
- Use data analytics to personalize marketing and promotional efforts. Create targeted offers based on guests' booking history, preferences, and behaviors. Send personalized promotions to past guests who stayed for shorter durations to entice them for longer stays.

### Recommended Strategies to increase Repeat bookings➖

Having uncovered repeat bookings pattern, I recommend the following:— 

- Implement targeted loyalty programs to encourage repeat bookings.

- Launch initiatives targeting the **Transient** customers, the Direct, Corporate and online TA/TO market segments, which comprises a significant portion of both old and new guests.

- Leverage insights on booking channels such as **TA/TO and Corporate** to optimize marketing strategies for both old and new guests.
- Further explore opportunities to attract guests from other regions, complementing the existing strong representation from Portugal and the United Kingdom.
- Focus on providing an exceptional guest experience to build loyalty. Satisfied guests are less likely to cancel reservations. Ensure that services, amenities, and facilities meet or exceed guest expectations.